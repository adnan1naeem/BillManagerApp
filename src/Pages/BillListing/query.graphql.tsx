import gql from 'graphql-tag'

export const query = gql`
query bills{
  allBills{
     name
  site
  month
  budget
  unitrate
  asset
  }
}
`