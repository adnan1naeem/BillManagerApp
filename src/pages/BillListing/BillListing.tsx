import React from 'react'
import { Text, ActivityIndicator } from "react-native";
import BillCard from "../../components/BillCard/Index"
import { styles } from "./style"
import { BillSubscription } from "./Subscriptionquery.graphql"
import { View } from 'react-native'


export interface Props {
	subscribeToMore: (event: any) => void,
	data: { allBills: any[] },
	loading?: boolean,
	empty?: boolean,
	Loading: boolean,
	error?: boolean
}

class BillListing extends React.PureComponent<Props, {}> {

	constructor(props: Props) {
		super(props);

	}

	componentDidMount() {
		this.props.subscribeToMore({
			document: BillSubscription,
			updateQuery: (prev: any, { subscriptionData }: any) => {

				if (!subscriptionData.data) return prev;

				if (subscriptionData.data && subscriptionData.data.Bill && subscriptionData.data.Bill.mutation === 'CREATED') {
					const newFeedItem = subscriptionData.data.Bill.node;

					const edge = {
						node: newFeedItem,
						_typename: 'Bill'
					}
					return Object.assign({}, prev, {
						allBills: [...prev.allBills, edge.node],
						_typename: 'Bill'
					});
				}

				if (subscriptionData.data && subscriptionData.data.Bill && subscriptionData.data.Bill.mutation === 'DELETED') {
					const newFeedItem = subscriptionData.data.Bill.previousValues;

					const FilteredList = prev.allBills.filter((data: { id: string }) => {
						return data.id != newFeedItem.id;
					});

					return Object.assign({}, prev, {
						allBills: FilteredList,
						_typename: 'Bill'
					});
				}

				if (subscriptionData.data && subscriptionData.data.Bill && subscriptionData.data.Bill.mutation === 'UPDATED') {

					const updatedBills = subscriptionData.data.Bill.node
					const bills = prev.allBills;

					const oldBillsIndex = bills.findIndex((bill: { id: string }) => {
						return updatedBills.id === bill.id
					})

					bills[oldBillsIndex] = updatedBills


					return {
						allBills: bills,
					}

				}
			}
		})

	}

	render() {
		if (!this.props.data.allBills && this.props.Loading == true || this.props.loading === true) {
			return (
				<View style={styles.IndicatorStyles}>
					<ActivityIndicator size={20} />
				</View>
			);
		}

		if (this.props.data && this.props.data.allBills && this.props.data.allBills.length < 0 && this.props.Loading === false || this.props.empty === true) {
			return <Text>No Data found</Text>
		}

		if (this.props.data && this.props.data.allBills && this.props.Loading == false && this.props.error != true) {
			return (

				<View style={styles.container}>
					{this.props.data.allBills.map((Data: any, i: any) =>
						<BillCard
							key={i}
							name={Data.name}
							month={Data.month}
							budget={Data.budget}
						/>)
					}
				</View>
			);
		}
		else {
			return <Text>Error fetching data</Text>;
		}
	}
}
export default BillListing;
