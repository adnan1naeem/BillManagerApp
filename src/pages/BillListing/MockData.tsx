export const Data = [{

  month: 'january',
  budget: 2000,
  name: 'ADNAN'
}]