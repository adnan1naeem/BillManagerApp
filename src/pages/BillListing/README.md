#  BillListing Component

## Compatibility

| 🌏 Web                   | 🖥 Electron              | 📱 React Native    |
| :----------------------: | :----------------------: | :----------------: |
| :heavy_multiplication_x: | :heavy_multiplication_x: | :heavy_check_mark: |

## Screenshots

| 🌏 Web | 🖥 Electron | 📱 React Native                           |
| :----: | :---------: | :---------------------------------------: |
| TBD    | TBD         | ![BILLLISTING](./Screenshots/Listing.png) |
|        |

## Props

| Name | Type   | Default | Description                               |
| :--- | :----- | :------ | :---------------------------------------- |
| data | object |         | It is used for adding  BillCards  in Page |

## How to use

```react
import React from 'react';
        <BillListing  data=[{name:'BillName' month="january" budget={2000} }]/>
```

## Tests
