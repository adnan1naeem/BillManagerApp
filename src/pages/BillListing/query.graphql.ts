import gql from 'graphql-tag'

export const query = gql`
query bills{
  allBills{
  id
  name
  site
  month
  budget
  unitrate
  asset
  }
}
`