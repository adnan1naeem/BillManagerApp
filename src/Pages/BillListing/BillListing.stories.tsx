import React from 'react';
import { storiesOf } from '@storybook/react-native';
import BillListing from './BillListing';
import { withKnobs, text, } from '@storybook/addon-knobs/react';
import { View } from 'react-native';
import { Query } from "react-apollo";
import gql from 'graphql-tag'
import { query } from './query.graphql'
import apolloStorybookDecorator from "apollo-storybook-react"
import { Data } from "./MockData"
import { allBills } from '../../components/BillCard/MockData';

storiesOf('BillListing', module)
    .addDecorator(getStory => <View>{getStory()}</View>)
    .addDecorator(withKnobs)
    .add('BillListing Page story', () => (
        <BillListing />
    ))


// export interface typedef {

//     allBills: [{ month: string, name: string, budget: number }],

// }

// const mocks = {

//     Query: Data
// }
// storiesOf('Apollo Client', module).addDecorator(
//     apolloStorybookDecorator({
//         typedef,
//         mocks,
//     })
// ).add('BillListing Page story', () => (
//     <BillListing />
// ))
