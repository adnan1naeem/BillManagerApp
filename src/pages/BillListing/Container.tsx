import React from 'react'
import { ScrollView, Text, ActivityIndicator, StatusBar } from "react-native";
import { Query } from 'react-apollo'
import { query } from "./query.graphql"
import BillListing from "./BillListing"

const Container = () => {

  return (
    <ScrollView>
      <Query query={query}>
        {({ data, loading, subscribeToMore }: any) => {
          return (<BillListing Loading={loading} data={loading === false ? data : {}} subscribeToMore={subscribeToMore} />);
        }}
      </Query>
    </ScrollView>
  );
}

export default Container;
