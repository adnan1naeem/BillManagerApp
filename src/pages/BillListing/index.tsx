import BillListing from  "./BillListing";
export default BillListing; 