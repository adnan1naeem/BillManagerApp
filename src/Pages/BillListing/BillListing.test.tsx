import * as React from 'react'
import { shallow, configure } from 'enzyme';
import BillListing from "./BillListing";
import ReactSixteenAdapter from 'enzyme-adapter-react-16';
import { Data } from "./MockData"
// it('renders without crashing', () => {
//   const rendered = renderer.create(<App />).toJSON()
//   expect(rendered).toMatchSnapshot()
//   expect(rendered).toBeTruthy()
// })
//configure({ adapter: new ReactSixteenAdapter() })


describe('BillCard Testcases', () => {

  it('should render component', () => {
    expect(<BillListing data={Data} />).toBeTruthy();
  });

  it('should throw error as required prop data is null', () => {
    let error
    try {
      shallow(<BillListing data={null} />)
    } catch (e) {
      error = e
    }
    expect(error).toBeDefined()
  });
  it('should throw error as required prop data is null', () => {
    let error
    try {
      shallow(<BillListing data={undefined} />)
    } catch (e) {
      error = e
    }
    expect(error).toBeDefined()
  });
});
