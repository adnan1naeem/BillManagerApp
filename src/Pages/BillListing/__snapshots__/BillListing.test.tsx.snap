// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "height": 200,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size={20}
  />
</View>
`;
